lib/elixir/lib/module/types/descr.ex

```diff
@@ -95,18 +135,33 @@ defmodule Module.Types.Descr do
   @compile {:inline, intersection: 3}
   defp intersection(:bitmap, v1, v2), do: bitmap_intersection(v1, v2)
   defp intersection(:atom, v1, v2), do: atom_intersection(v1, v2)
+  defp intersection(:dynamic, v1, v2), do: dynamic_intersection(v1, v2)
 
   @doc """
   Computes the difference between two types.
   """
   def difference(left = %{}, right = %{}) do
+    if gradual?(left) or gradual?(right) do
+      {left_dynamic, left_static} = Map.pop(left, :dynamic, left)
+      {right_dynamic, right_static} = Map.pop(right, :dynamic, right)
+
+      difference_static(left_static, right_dynamic)
+      |> Map.put(:dynamic, difference_static(left_dynamic, right_static))
+    else
+      difference_static(left, right)
+    end
+  end
+
+  # For static types, the difference is component-wise.
+  defp difference_static(left, right) do
     iterator_difference(:maps.next(:maps.iterator(right)), left)
   end
 
   # Returning 0 from the callback is taken as none() for that subtype.
   @compile {:inline, difference: 3}
   defp difference(:bitmap, v1, v2), do: bitmap_difference(v1, v2)
   defp difference(:atom, v1, v2), do: atom_difference(v1, v2)
+  defp difference(:dynamic, v1, v2), do: dynamic_difference(v1, v2)
```

Comment:
We don't ever use this clause, do we? That's why we have the bug @sabiwara just shared. We never check if `dynamic_difference` returns `empty?` and ignore its result if we do.

We should probably add tests for union/intersection/difference that shows we discard the dynamic once it is empty.

Reply:
@sabiwara I second this, it was a bug (thank you for catching it!). During intersections the field is removed if dynamic is empty thanks to `dynamic_intersection`, but for differences I had to modify the body of `difference` (`dynamic_difference` is only used for subtyping). With the dynamic/static inequality invariant, if the dynamic part is empty then the entire type is `@none`.

ps: added tests for difference and intersection. for union, the dynamic part can only grow and not be deleted

Reply:
Should we remove line as we will never reach this?

```suggestion
```

Reply:
It's reached when we compute subtyping on two gradual types with a defined :dynamic field. In that case checking for the emptiness of the component-wise difference (`difference_static`) is the subtype check.

Reply:
Ah, in the sub typing, I missed that, thanks!